Collate tpms 33

mikej888 opened the pull request:
Refactor collate_tpms.R #33 updates:

* Conforms to R styles.
* File and function comments in Roxygen2 format.
* Fixed bug with how ORF lists are compared.
* ORFs are sorted upon loading from an ORF FASTA file or a sample-specific TPMs TSV file. TPMs from sample-specific TPMs files are sorted by ORF also. This ensures that collation can occur regardless of the order of TPMs in individual TPMs files or in the ORF FASTA file.
* `--tpms-file` is assumed to be a relative or absolute path i.e. it is no longer relative to `--output-dir`.
* `<sample_name> <file_name>` pairs are expected at the command-line. This removes the need for `--output-dir` and `--samples-subdir` command-line parameters and simplifies the code as directory or file names no longer need to be parsed to infer sample names.
* Extracted all non-command-line-related code from `collate_tpms.R` to `collate_tpms_functions.R`.
* `test_collate_tpms.R` includes testthat-compliant tests. These create their own sample-specific TPMs files and a FASTA file (for testing `--orf_fasta`).

ewallace commented:
I reviewed this pull request and ran the tests. The documentation and tests look great and the functions are also readable and work.

I put in two commits:
* made `digits` an optional argument of `CollateTpms` so that if the user wants different precision output they can; also removes need for separate function definition.
* made `sort_orfs` an optional argument where possible in `collate_tpms_functions.R`.

More about sorting the ORFs: from my perspective as a user I don't want to do that. Usually I have my ORFs in a biologically relevant order (e.g. genomic position) in the fasta file and so reordering them is undesirable. Also I don't think this will be a problem in riboviz because AFAIK we maintain the input order in the FASTA file. To double-check, the vignette inputs should start with YAL069W at the top both for fasta and for the sample-specific TPMs. We also have the `all.equal` test that will raise an error if the values are differently ordered, so I feel it is safe that will catch problems.

So, in commit dbc1dd5 I made`sort_orfs` optional in `collate_tpms_functions.R`, keeping default value TRUE. I would prefer giving it default value `FALSE`, but I think that would break some tests so I did not try to fix that myself.
 
 It is not important to me if we pass in `sort_orfs` or `digits` as command-line options to `collate_tpms.R`, but I appreciate that the code would be neater and more complete if that were the case.
 
 Lastly I think it would be nice to have the flexibility to give `collate_tpms.R`
 * EITHER (as currently) a paired list of sample names and filenames
 * OR (more like [featureCounts from subread](http://subread.sourceforge.net/)) just a list of files. Then the column names in the output would be given by the filenames.

This is a nice-to-have and not essential, so it could be put in a new issue ticket, or if there is some reason it is a bad idea it does not need to happen.

After addressing the `sort_orfs` question, let's merge this in. Given it passes unit tests and regression tests, we can pick up any unexpected issues in integration tests.

mikej888 commented:
> Also I don't think this will be a problem in riboviz because AFAIK we maintain the input order in the FASTA file.
> To double-check, the vignette inputs should start with YAL069W at the top both for fasta and for the sample-specific TPMs.

This is not caused by`collate_tpms.R`. See new issue Update bam_to_h5.R so it does not reorder ORFs #360.

mikej888 commented:
> Lastly I think it would be nice to have the flexibility to give collate_tpms.R
>
> EITHER (as currently) a paired list of sample names and filenames
> OR (more like featureCounts from subread) just a list of files. Then the column names in the output would be given by the filenames.
> This is a nice-to-have and not essential, so it could be put in a new issue ticket, or if there is some reason it is a bad idea it does not need to happen.

Raised Allow collate_tpms.R to take a list of filenames only #359.

mikej888 commented:
As of commit 375cec7:

* Added `ignore_row_order` (default `False`) to `riboviz.utils.equal_tsv` and `equal_dataframes`.
* Updated `riboviz.test.regression.test_collate_tpms_tsv` to call `utils.equal_tsv` with `ignore_row_order=True`
* Added `sort_orfs` and `digits` command-line parameters to `collate_tpms.R`. Set `sort_orfs` to `FALSE` by default in `collate_tpms.R` and `collate_tpms_functions.R`.
* Extended `test_collate_tpms.R`:
  - `--orf-fasta` test replaced by `--orf-fasta (sort_orfs=TRUE)` and `--orf-fasta (sort_orfs=FALSE)` tests.
  - `Sample files with ORFs in different orders` replaced by `sort_orfs=TRUE` and `sort_orfs=FALSE` tests.
  - Added `digits=0` test.
